### PR TITLE
feat(op_crates/web): Add all single byte encodings to TextDecoder

### PR DESCRIPTION
**Pull request by AliBasicCoder**
i added this ecodings

```
    ibm866
    iso-8859-2
    iso-8859-3
    iso-8859-4
    iso-8859-5
    iso-8859-6
    iso-8859-7
    iso-8859-8
    iso-8859-10
    iso-8859-13
    iso-8859-14
    iso-8859-15
    iso-8859-16
    koi8-r
    koi8-u
    macintosh
    windows-874
    windows-1250
    windows-1251
    windows-1253
    windows-1254
    windows-1255
    windows-1256
    windows-1257
    windows-1258
    x-mac-cyrillic
```

NOTES:
  * i needed to make code points start from 0 not 128
  * i made code points 8 items per line

### CODE REVIEW

**Comment by ry** [comment]
@AliBasicCoder Thanks for the patch. I haven't looked at this in detail yet - but this needs tests before it can be landed.

**Comment by kitsonk** [comment]
Just to point out, browsers only support encoding UTF-8. https://encoding.spec.whatwg.org/#interface-textencoder so this would be a significant variation from the spec.

**Comment by lucacasonato** [comment]
> Just to point out, browsers only support encoding UTF-8. https://encoding.spec.whatwg.org/#interface-textencoder so this would be a significant variation from the spec.

@kitsonk This is about decoding though. Encoding should stay utf-8 only to be spec compliant, but decoder can have many encodings (Chrome has many encodings for decoding, but only utf-8 for encoding).

**Comment by AliBasicCoder** [comment]
> Just to point out, browsers only support encoding UTF-8. https://encoding.spec.whatwg.org/#interface-textencoder so this would be a significant variation from the spec.

i just added for decoding and didn't touch TextEncoder at all

**Comment by kitsonk** [comment]
🤦 I can't read.  Sorry folks.

**Comment by AliBasicCoder** [comment]
i have just added all the labels for the encodings

**Comment by ry** [comment]
@AliBasicCoder How about in `cli/tests/unit/text_encoding_test.ts`? Or maybe create a new test module neighboring that one. (If you create a new one, be sure to add it to cli/tests/unit/unit_tests.ts)

Tests can be run with `cargo test unit`

**Comment by AliBasicCoder** [comment]
> @AliBasicCoder How about in `cli/tests/unit/text_encoding_test.ts`? Or maybe create a new test module neighboring that one. (If you create a new one, be sure to add it to cli/tests/unit/unit_tests.ts)
> 
> Tests can be run with `cargo test unit`

test are added now

File: cli/tests/unit/single_byte_encoding_test.ts [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+
+// function unitTest(func: Function) {
+//   try {
+//     func();
+//     console.log(`(success) test ${func.name}`);
+//   } catch (error) {
+//     // prettier-ignore
+//     console.log(`(fail)    test ${func.name}${" ".repeat(12 - func.name.length)}  ${error.message}`);
+//   }
+// }
```
Comment (ry) [inline cli/tests/unit/single_byte_encoding_test.ts:19]: Remove commented out code.

File: cli/tests/unit/single_byte_encoding_test.ts [hunk 2]
```diff
@@ -0,0 +1,482 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
```
Comment (ry) [inline cli/tests/unit/single_byte_encoding_test.ts:1]: Maintain copyright line of WPT

File: cli/tests/unit/single_byte_encoding_test.ts [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+          const data = decoder.decode(view);
+
+          assertEquals(decoder.encoding, encoding.name.toLowerCase());
+          // @ts-ignore
+          assertDecode(data, encoding.name);
+        },
+      }[label]
+    );
+  }
+}
```
Comment (ry) [inline cli/tests/unit/single_byte_encoding_test.ts:482]: 👍

**Comment by ry** [comment]
@AliBasicCoder We are missing a CLA from you. I believe you've signed it before - it's the same terms - but we updated it changing my personal name to an entity Deno Land Inc.

**Comment by bartlomieju** [comment]
@AliBasicCoder I'm interested in landing this PR, could you rebase it and move the logic to `op_crates/web/08_text_encoding.js`?

**Comment by AliBasicCoder** [comment]
> @AliBasicCoder I'm interested in landing this PR, could you rebase it and move the logic to `op_crates/web/08_text_encoding.js`?

done

**Comment by AliBasicCoder** [comment]
the problem is 
```git status -uno --porcelain --ignore-submodules ```
prints
```
 M cli/tests/unit/single_byte_encoding_test.ts
 M op_crates/web/08_text_encoding.js
 M op_crates/web/text_encoding_test.js
```

File: op_crates/web/08_text_encoding.js [hunk 4]
```diff
@@ -207,10 +344,2915 @@
...
+    143,
+    144,
+    8216,
+    8217,
+    8220,
+    8221,
+    8226,
+    8211,
+    8212,
+    732,
```
Comment (bartlomieju) [inline op_crates/web/08_text_encoding.js:376]: Note to self: fix formatting before landing. 6 or 8 elements per line

**Review by bartlomieju — APPROVED** [review]
LGTM. Huge thanks to @AliBasicCoder and sorry it took so long to land

**Comment by linux-china** [comment]
@AliBasicCoder   any plan to add utf-16le?   Now emscripten's  js code use utf-16le for decoding.   I want to run wasm compiled by emscripten, and  utf-16le is a problem.  if you can implement it, and it will be very greatful to run emscripten wasm on Deno.  thanks.  

```
// Given a pointer 'ptr' to a null-terminated UTF16LE-encoded string in the emscripten HEAP, returns
// a copy of that string as a Javascript String object.

var UTF16Decoder = typeof TextDecoder !== 'undefined' ? new TextDecoder('utf-16le') : undefined;

```

**Comment by AliBasicCoder** [comment]
> @AliBasicCoder   any plan to add utf-16le?   Now emscripten's  js code use utf-16le for decoding.   I want to run wasm compiled by emscripten, and  utf-16le is a problem.  if you can implement it, and it will be very greatful to run emscripten wasm on Deno.  thanks.  
> 
> ```
> // Given a pointer 'ptr' to a null-terminated UTF16LE-encoded string in the emscripten HEAP, returns
> // a copy of that string as a Javascript String object.
> 
> var UTF16Decoder = typeof TextDecoder !== 'undefined' ? new TextDecoder('utf-16le') : undefined;
> 
> ```

see #8108

**Comment by xaliphostes** [comment]
> @AliBasicCoder any plan to add utf-16le? Now emscripten's js code use utf-16le for decoding. I want to run wasm compiled by emscripten, and utf-16le is a problem. if you can implement it, and it will be very greatful to run emscripten wasm on Deno. thanks.
> 
> ```
> // Given a pointer 'ptr' to a null-terminated UTF16LE-encoded string in the emscripten HEAP, returns
> // a copy of that string as a Javascript String object.
> 
> var UTF16Decoder = typeof TextDecoder !== 'undefined' ? new TextDecoder('utf-16le') : undefined;
> ```

Agreed
Same problem while using wasm with deno.

**Pull request merged**